pcdsdevices/signal.py

```diff
@@ -428,23 +428,72 @@ class PVStateSignal(AggregateSignal):
 
     See `AggregateSignal` for more information.
     """
+    _metadata_keys = Signal._core_metadata_keys + (
+        'enum_strs',
+    )
 
     def __init__(self, *, name, **kwargs):
         super().__init__(name=name, **kwargs)
         for attr_name in self.parent._state_logic:
             self.add_signal_by_attr_name(attr_name)
+        self._has_setpoint_md = False
 
-    def describe(self):
+    @property
+    def enum_strs(self) -> Tuple[str, ...]:
+        """
+        Mimic the epics signal property for enum strings.
+        """
+        return tuple(state.name for state in self.parent.states_enum)
+
+    def describe(self) -> Dict[str, Dict[str, Any]]:
+        """
+        Make sure a reasonable description exists for bluesky scans.
+        """
         # Base description information
         sub_sigs = [sig.name for sig in self._signals]
-        desc = {'source': 'SUM:{}'.format(','.join(sub_sigs)),
-                'dtype': 'string',
-                'shape': [],
-                'enum_strs': tuple(state.name
-                                   for state in self.parent.states_enum)}
+        desc = {
+            'source': 'SUM:{}'.format(','.join(sub_sigs)),
+            'dtype': 'string',
+            'shape': [],
+            'enum_strs': self.enum_strs,
+        }
         return {self.name: desc}
 
-    def _calc_readback(self):
+    def _calc_readback(self) -> str:
+        """
+        Implements the PVStatePositioner logic.
+
+        This relies on this signal being a component of a PVStatePositioner
+        class with properly-defined class attributes.
+
+        On the first call, we'll do some setup of metadata values, since
+        this first call happens after the signals connect and we're guaranteed
+        that everything is ready. Doing it earlier can run into some
+        race conditions.
+        """
+        # Do some one-time setup here
```

Comment:
I'm not 100% sure if this was the right place to do this

Reply:
The one-time setup in the first callback seems appropriate and efficient to me.

The only thing worth noting is that `_calc_readback` happens inside of an acquired [RLock](https://github.com/klauer/pcdsdevices/blob/408b62f2baf6b8d735609a018629cb7d143cc07f/pcdsdevices/signal.py#L197-L216) . I see nothing in the block that points to a potential deadlock condition though: running subscriptions happens in a separate thread and shouldn't block.

Reply:
Since it is in an `RLock` though I might take a moment to see if I can set it up earlier in the chain, both to avoid potential mishaps and to not set a bad example for future people

Reply:
Can't quickly find a better way, I'm going to move on